File: packages/contracts/python/demodirector_contracts/longform.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, model_validator

from demodirector_contracts.models import ContractModel, NonEmptyString, Viewport

LongFormSectionId = Literal[
    "hook",
    "problem",
    "promise",
    "product_walkthrough",
    "trust_technology",
    "editing_control",
    "finished_result",
    "closing",
]
LongFormStepKind = Literal["narrative", "template", "attention", "style"]

SECTION_PROFILE: tuple[tuple[LongFormSectionId, int], ...] = (
    ("hook", 4_000),
    ("problem", 6_000),
    ("promise", 10_000),
    ("product_walkthrough", 75_000),
    ("trust_technology", 35_000),
    ("editing_control", 25_000),
    ("finished_result", 20_000),
    ("closing", 5_000),
)


class LongFormDirectionRequest(ContractModel):
    plan_id: NonEmptyString
    run_id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    parent_motion_run_id: NonEmptyString
    visual_variant: Literal[
        "editorial_story", "product_spotlight", "technical_proof"
    ]
    research_status: Literal["complete", "degraded"]
    evidence_refs: list[NonEmptyString] = Field(min_length=1, max_length=160)
    parallel_source_refs: list[NonEmptyString] = Field(default_factory=list, max_length=80)
    product_clip_refs: list[NonEmptyString] = Field(min_length=1, max_length=80)
    target_ids: list[NonEmptyString] = Field(default_factory=list, max_length=160)
    created_at: datetime


class NarrativeSection(ContractModel):
    id: LongFormSectionId
    order: int = Field(ge=0, le=7)
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    narration: str = Field(min_length=1, max_length=1_800)
    source_refs: list[NonEmptyString] = Field(min_length=1, max_length=40)
    product_clip_ref: NonEmptyString
    product_visible: Literal[True] = True


class VisualBeat(ContractModel):
    id: NonEmptyString
    section_id: LongFormSectionId
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    purpose: Literal[
        "hook",
        "problem",
        "promise",
        "product_operation",
        "create_action",
        "finished_glimpse",
        "walkthrough",
        "evidence",
        "control",
        "result",
        "closing",
    ]
    template_id: Literal[
        "hook",
        "framed_product",
        "feature_callout",
        "split_explanation",
        "proof_safety",
        "closing",
    ]
    product_clip_ref: NonEmptyString
    target_id: NonEmptyString | None = None
    source_refs: list[NonEmptyString] = Field(default_factory=list, max_length=20)


class CaptureChapter(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    order: int = Field(ge=0, le=7)
    section_ids: list[LongFormSectionId] = Field(min_length=1, max_length=4)
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    viewport: Viewport
    account_ref: NonEmptyString
    project_state_ref: NonEmptyString
    style_version: Literal["style-v1"] = "style-v1"
    cursor_continuity_key: NonEmptyString


class ChapterCheckpoint(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    plan_id: NonEmptyString
    chapter_id: NonEmptyString
    status: Literal["pending", "captured", "failed", "approved"]
    attempt: int = Field(ge=0, le=3)
    artifact_ref: NonEmptyString | None = None
    artifact_sha256: str | None = Field(default=None, pattern=r"^[a-f0-9]{64}$")
    elapsed_ms: int | None = Field(default=None, ge=0)
    replaces_checkpoint_id: NonEmptyString | None = None
    created_at: datetime

    @model_validator(mode="after")
    def artifact_matches_status(self) -> ChapterCheckpoint:
        has_artifact = self.artifact_ref is not None and self.artifact_sha256 is not None
        if self.status in {"captured", "approved"} and not has_artifact:
            raise ValueError("Captured chapters require a validated artifact")
        if self.status in {"pending", "failed"} and has_artifact:
            raise ValueError("Incomplete chapters cannot claim a validated artifact")
        return self


class CondensedInterval(ContractModel):
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    actual_elapsed_ms: int = Field(gt=0)
    label: NonEmptyString = "Time compressed"

    @model_validator(mode="after")
    def preserves_measurement(self) -> CondensedInterval:
        if self.end_ms <= self.start_ms:
            raise ValueError("Condensed interval must have positive display duration")
        if self.actual_elapsed_ms <= self.end_ms - self.start_ms:
            raise ValueError("Actual elapsed time must exceed the condensed display interval")
        return self


class AudioMixPlan(ContractModel):
    narration_lufs: float = Field(default=-16, ge=-24, le=-12)
    music_lufs: float = Field(default=-28, ge=-40, le=-20)
    ducking_db: float = Field(default=-8, ge=-18, le=-3)
    fade_ms: int = Field(default=500, ge=100, le=2_000)


class LongFormVideoPlan(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    adk_run_id: NonEmptyString
    parent_motion_run_id: NonEmptyString
    version: Literal["longform-v1"] = "longform-v1"
    motion_version: Literal["motion-v1"] = "motion-v1"
    style_version: Literal["style-v1"] = "style-v1"
    visual_variant: Literal[
        "editorial_story", "product_spotlight", "technical_proof"
    ]
    duration_ms: Literal[180000] = 180_000
    research_status: Literal["complete", "degraded"]
    evidence_refs: list[NonEmptyString] = Field(min_length=1, max_length=160)
    parallel_source_refs: list[NonEmptyString] = Field(default_factory=list, max_length=80)
    product_clip_refs: list[NonEmptyString] = Field(min_length=1, max_length=80)
    target_ids: list[NonEmptyString] = Field(default_factory=list, max_length=160)
    sections: list[NarrativeSection] = Field(min_length=8, max_length=8)
    beats: list[VisualBeat] = Field(min_length=20, max_length=30)
    chapters: list[CaptureChapter] = Field(min_length=2, max_length=8)
    condensed_intervals: list[CondensedInterval] = Field(default_factory=list, max_length=12)
    audio_mix: AudioMixPlan = Field(default_factory=AudioMixPlan)
    first_product_operation_ms: int = Field(ge=0, le=20_000)
    create_action_ms: int = Field(ge=0, le=30_000)
    finished_glimpse_ms: int = Field(ge=0, le=45_000)
    product_presence_percent: float = Field(ge=90, le=100)
    created_at: datetime

    @model_validator(mode="after")
    def proof_first_structure_is_valid(self) -> LongFormVideoPlan:
        expected_ids = [item[0] for item in SECTION_PROFILE]
        expected_durations = [item[1] for item in SECTION_PROFILE]
        if [item.id for item in self.sections] != expected_ids:
            raise ValueError("Long-form sections must use the fixed narrative order")
        cursor = 0
        for index, section in enumerate(self.sections):
            if section.order != index or section.start_ms != cursor:
                raise ValueError("Long-form sections must be contiguous")
            if section.end_ms - section.start_ms != expected_durations[index]:
                raise ValueError("Long-form section duration does not match the profile")
            cursor = section.end_ms
        if cursor != self.duration_ms:
            raise ValueError("Long-form sections must total exactly 180 seconds")
        words = sum(len(section.narration.split()) for section in self.sections)
        if not 350 <= words <= 430:
            raise ValueError("Long-form narration must contain 350 to 430 words")
        allowed_evidence = set(self.evidence_refs)
        allowed_clips = set(self.product_clip_refs)
        allowed_targets = set(self.target_ids)
        if self.research_status == "complete" and not self.parallel_source_refs:
            raise ValueError("Complete research requires direct Parallel evidence")
        if set(self.parallel_source_refs) - allowed_evidence:
            raise ValueError("Parallel references must be saved project evidence")
        for section in self.sections:
            section_words = len(section.narration.split())
            section_seconds = (section.end_ms - section.start_ms) / 1_000
            if section_words > section_seconds * 3:
                raise ValueError("Narration exceeds the readable speech rate for its section")
            if set(section.source_refs) - allowed_evidence:
                raise ValueError("Narrative sections must cite saved project evidence")
            if section.product_clip_ref not in allowed_clips:
                raise ValueError("Narrative sections must use approved product clips")
        section_by_id = {section.id: section for section in self.sections}
        beat_sections: set[LongFormSectionId] = set()
        for beat in self.beats:
            section = section_by_id[beat.section_id]
            beat_sections.add(beat.section_id)
            if beat.start_ms < section.start_ms or beat.end_ms > section.end_ms:
                raise ValueError("Visual beats must fit their narrative section")
            if beat.end_ms <= beat.start_ms or beat.product_clip_ref not in allowed_clips:
                raise ValueError("Visual beats must use valid time and product footage")
            if beat.target_id is not None and beat.target_id not in allowed_targets:
                raise ValueError("Visual beats must use approved attention targets")
            if set(beat.source_refs) - allowed_evidence:
                raise ValueError("Visual beats must cite saved project evidence")
        if beat_sections != set(expected_ids):
            raise ValueError("Every narrative section requires a visual beat")
        milestone_beats = {
            purpose: [beat for beat in self.beats if beat.purpose == purpose]
            for purpose in ("product_operation", "create_action", "finished_glimpse")
        }
        if any(len(beats) != 1 for beats in milestone_beats.values()):
            raise ValueError("Proof-first milestones require exactly one visual beat")
        if (
            milestone_beats["product_operation"][0].start_ms
            != self.first_product_operation_ms
            or milestone_beats["create_action"][0].start_ms != self.create_action_ms
            or milestone_beats["finished_glimpse"][0].start_ms
            != self.finished_glimpse_ms
        ):
            raise ValueError("Proof-first milestones require matching visual beats")
        chapter_sections = [item for chapter in self.chapters for item in chapter.section_ids]
        if chapter_sections != expected_ids:
            raise ValueError("Capture chapters must cover each section once in order")
        first_chapter = self.chapters[0]
        continuity = (
            first_chapter.viewport,
            first_chapter.account_ref,
            first_chapter.project_state_ref,
            first_chapter.style_version,
            first_chapter.cursor_continuity_key,
        )
        for index, chapter in enumerate(self.chapters):
            chapter_sections_data = [section_by_id[item] for item in chapter.section_ids]
            if (
                chapter.project_id != self.project_id
                or chapter.order != index
                or chapter.start_ms != chapter_sections_data[0].start_ms
                or chapter.end_ms != chapter_sections_data[-1].end_ms
                or chapter.end_ms <= chapter.start_ms
            ):
                raise ValueError("Capture chapter lineage or timing is invalid")
            if (
                chapter.viewport,
                chapter.account_ref,
                chapter.project_state_ref,
                chapter.style_version,
                chapter.cursor_continuity_key,
            ) != continuity:
                raise ValueError("Capture chapters must preserve recording continuity")
        previous_end = 0
        for interval in self.condensed_intervals:
            if interval.start_ms < previous_end or interval.end_ms > self.duration_ms:
                raise ValueError("Condensed intervals must be ordered within the video")
            previous_end = interval.end_ms
        return self

    def narration_word_count(self) -> int:
        return sum(len(section.narration.split()) for section in self.sections)


class LongFormADKStep(ContractModel):
    id: NonEmptyString
    run_id: NonEmptyString
    project_id: NonEmptyString
    kind: LongFormStepKind
    status: Literal["succeeded", "failed"]
    tool_call_count: int = Field(ge=0, le=1)
    artifact_refs: list[NonEmptyString] = Field(default_factory=list, max_length=160)
    output_sha256: str | None = Field(default=None, pattern=r"^[a-f0-9]{64}$")
    created_at: datetime


class LongFormADKRun(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    parent_motion_run_id: NonEmptyString
    session_id: NonEmptyString
    agent_name: Literal["demodirector_longform_director"]
    model: NonEmptyString
    status: Literal["running", "succeeded", "failed"]
    runner_completed: bool
    workflow_runs: int = Field(ge=0, le=1)
    step_ids: list[NonEmptyString] = Field(default_factory=list, max_length=4)
    output_plan_id: NonEmptyString | None = None
    started_at: datetime
    completed_at: datetime | None = None
    elapsed_ms: int | None = Field(default=None, ge=0)
    message: str = Field(min_length=1, max_length=300)

    @model_validator(mode="after")
    def successful_run_is_complete(self) -> LongFormADKRun:
        if self.status == "succeeded" and (
            not self.runner_completed
            or self.workflow_runs != 1
            or len(self.step_ids) != 4
            or self.output_plan_id is None
            or self.completed_at is None
        ):
            raise ValueError("Successful long-form runs require four completed ADK steps")
        if self.status == "failed" and (self.runner_completed or self.output_plan_id):
            raise ValueError("Failed ADK runs cannot claim consumed output")
        return self


class LongFormValidationReport(ContractModel):
    project_id: NonEmptyString
    plan_id: NonEmptyString
    duration_ms: Literal[180000]
    width: Literal[2560]
    height: Literal[1440]
    fps: Literal[30]
    narration_word_count: int = Field(ge=350, le=430)
    visual_beat_count: int = Field(ge=20, le=30)
    product_presence_percent: float = Field(ge=90, le=100)
    has_audio: bool
    has_captions: bool
    synchronized: bool
    issues: list[NonEmptyString] = Field(default_factory=list)
    passed: bool

    @model_validator(mode="after")
    def passed_matches_checks(self) -> LongFormValidationReport:
        expected = self.has_audio and self.has_captions and self.synchronized and not self.issues
        if self.passed != expected:
            raise ValueError("Long-form validation status must match its checks")
        return self


class CompiledLongFormPlan(ContractModel):
    project_id: NonEmptyString
    plan_id: NonEmptyString
    chapter_count: int = Field(ge=2, le=8)
    beat_count: int = Field(ge=20, le=30)
    deterministic_hash: str = Field(pattern=r"^[a-f0-9]{64}$")

File: packages/contracts/src/longform.ts
import { z } from "zod";
import { viewportSchema } from "./schemas";
import { visualVariantIdSchema } from "./style";

const id = z.string().min(1);
export const longFormSectionIdSchema = z.enum([
  "hook", "problem", "promise", "product_walkthrough", "trust_technology",
  "editing_control", "finished_result", "closing",
]);
export const longFormDirectionRequestSchema = z.object({
  plan_id: id, run_id: id, project_id: id, job_id: id, parent_motion_run_id: id,
  visual_variant: visualVariantIdSchema, research_status: z.enum(["complete", "degraded"]),
  evidence_refs: z.array(id).min(1).max(160), parallel_source_refs: z.array(id).max(80).default([]),
  product_clip_refs: z.array(id).min(1).max(80), target_ids: z.array(id).max(160).default([]),
  created_at: z.string().datetime({ offset: true }),
}).strict();
const templateId = z.enum([
  "hook", "framed_product", "feature_callout", "split_explanation", "proof_safety", "closing",
]);
export const longFormNarrativeSectionSchema = z.object({
  id: longFormSectionIdSchema, order: z.number().int().min(0).max(7),
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
  narration: z.string().min(1).max(1800), source_refs: z.array(id).min(1).max(40),
  product_clip_ref: id, product_visible: z.literal(true).default(true),
}).strict();
export const visualBeatSchema = z.object({
  id, section_id: longFormSectionIdSchema,
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
  purpose: z.enum(["hook", "problem", "promise", "product_operation", "create_action", "finished_glimpse", "walkthrough", "evidence", "control", "result", "closing"]),
  template_id: templateId, product_clip_ref: id, target_id: id.nullable().default(null),
  source_refs: z.array(id).max(20).default([]),
}).strict();
export const captureChapterSchema = z.object({
  id, project_id: id, order: z.number().int().min(0).max(7),
  section_ids: z.array(longFormSectionIdSchema).min(1).max(4),
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
  viewport: viewportSchema, account_ref: id, project_state_ref: id,
  style_version: z.literal("style-v1").default("style-v1"), cursor_continuity_key: id,
}).strict();
export const chapterCheckpointSchema = z.object({
  id, project_id: id, plan_id: id, chapter_id: id,
  status: z.enum(["pending", "captured", "failed", "approved"]),
  attempt: z.number().int().min(0).max(3), artifact_ref: id.nullable().default(null),
  artifact_sha256: z.string().regex(/^[a-f0-9]{64}$/).nullable().default(null),
  elapsed_ms: z.number().int().nonnegative().nullable().default(null),
  replaces_checkpoint_id: id.nullable().default(null), created_at: z.string().datetime({ offset: true }),
}).strict().superRefine((item, context) => {
  const hasArtifact = item.artifact_ref !== null && item.artifact_sha256 !== null;
  if (["captured", "approved"].includes(item.status) !== hasArtifact) {
    context.addIssue({ code: "custom", message: "Chapter artifact must match status" });
  }
});
export const chapterCheckpointsSchema = z.array(chapterCheckpointSchema);
export const condensedIntervalSchema = z.object({
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
  actual_elapsed_ms: z.number().int().positive(), label: id.default("Time compressed"),
}).strict().refine((item) => item.end_ms > item.start_ms && item.actual_elapsed_ms > item.end_ms - item.start_ms);
export const audioMixPlanSchema = z.object({
  narration_lufs: z.number().min(-24).max(-12).default(-16),
  music_lufs: z.number().min(-40).max(-20).default(-28),
  ducking_db: z.number().min(-18).max(-3).default(-8),
  fade_ms: z.number().int().min(100).max(2000).default(500),
}).strict();
export const longFormVideoPlanSchema = z.object({
  id, project_id: id, job_id: id, adk_run_id: id, parent_motion_run_id: id,
  version: z.literal("longform-v1").default("longform-v1"),
  motion_version: z.literal("motion-v1").default("motion-v1"),
  style_version: z.literal("style-v1").default("style-v1"),
  visual_variant: visualVariantIdSchema, duration_ms: z.literal(180000).default(180000),
  research_status: z.enum(["complete", "degraded"]), evidence_refs: z.array(id).min(1).max(160),
  parallel_source_refs: z.array(id).max(80).default([]), product_clip_refs: z.array(id).min(1).max(80),
  target_ids: z.array(id).max(160).default([]), sections: z.array(longFormNarrativeSectionSchema).length(8),
  beats: z.array(visualBeatSchema).min(20).max(30), chapters: z.array(captureChapterSchema).min(2).max(8),
  condensed_intervals: z.array(condensedIntervalSchema).max(12).default([]), audio_mix: audioMixPlanSchema.default({ narration_lufs: -16, music_lufs: -28, ducking_db: -8, fade_ms: 500 }),
  first_product_operation_ms: z.number().int().min(0).max(20000), create_action_ms: z.number().int().min(0).max(30000),
  finished_glimpse_ms: z.number().int().min(0).max(45000), product_presence_percent: z.number().min(90).max(100),
  created_at: z.string().datetime({ offset: true }),
}).strict().superRefine((plan, context) => {
  const profile = [["hook", 4000], ["problem", 6000], ["promise", 10000], ["product_walkthrough", 75000], ["trust_technology", 35000], ["editing_control", 25000], ["finished_result", 20000], ["closing", 5000]] as const;
  let cursor = 0;
  profile.forEach(([name, duration], index) => {
    const section = plan.sections[index];
    if (!section || section.id !== name || section.order !== index || section.start_ms !== cursor || section.end_ms - section.start_ms !== duration) context.addIssue({ code: "custom", message: "Invalid long-form timing profile", path: ["sections", index] });
    cursor += duration;
  });
  const words = plan.sections.flatMap((section) => section.narration.trim().split(/\s+/)).length;
  if (words < 350 || words > 430) context.addIssue({ code: "custom", message: "Narration must contain 350 to 430 words" });
  const evidence = new Set(plan.evidence_refs), clips = new Set(plan.product_clip_refs), targets = new Set(plan.target_ids);
  if (plan.research_status === "complete" && plan.parallel_source_refs.length === 0) context.addIssue({ code: "custom", message: "Complete research requires direct Parallel evidence" });
  if (plan.parallel_source_refs.some((ref) => !evidence.has(ref))) context.addIssue({ code: "custom", message: "Parallel evidence is not approved" });
  plan.sections.forEach((section) => {
    const sectionWords = section.narration.trim().split(/\s+/).length;
    if (sectionWords > ((section.end_ms - section.start_ms) / 1000) * 3) context.addIssue({ code: "custom", message: "Narration exceeds the readable speech rate for its section" });
    if (!clips.has(section.product_clip_ref) || section.source_refs.some((ref) => !evidence.has(ref))) context.addIssue({ code: "custom", message: "Section references are not approved" });
  });
  plan.beats.forEach((beat) => {
    const section = plan.sections.find((item) => item.id === beat.section_id);
    if (!section || beat.start_ms < section.start_ms || beat.end_ms > section.end_ms || beat.end_ms <= beat.start_ms || !clips.has(beat.product_clip_ref) || (beat.target_id !== null && !targets.has(beat.target_id)) || beat.source_refs.some((ref) => !evidence.has(ref))) context.addIssue({ code: "custom", message: "Beat timing or references are not approved" });
  });
  if (new Set(plan.beats.map((beat) => beat.section_id)).size !== profile.length) context.addIssue({ code: "custom", message: "Every narrative section requires a visual beat" });
  const milestones = [["product_operation", plan.first_product_operation_ms], ["create_action", plan.create_action_ms], ["finished_glimpse", plan.finished_glimpse_ms]] as const;
  milestones.forEach(([purpose, start]) => {
    const matching = plan.beats.filter((beat) => beat.purpose === purpose);
    if (matching.length !== 1 || matching[0]?.start_ms !== start) context.addIssue({ code: "custom", message: "Proof-first milestone beat is invalid" });
  });
  const sectionIds = plan.chapters.flatMap((chapter) => chapter.section_ids);
  if (sectionIds.join("|") !== profile.map(([name]) => name).join("|")) context.addIssue({ code: "custom", message: "Capture chapters must cover each section once in order" });
  const continuity = plan.chapters[0] ? JSON.stringify([plan.chapters[0].viewport, plan.chapters[0].account_ref, plan.chapters[0].project_state_ref, plan.chapters[0].style_version, plan.chapters[0].cursor_continuity_key]) : "";
  plan.chapters.forEach((chapter, index) => {
    const chapterSections = chapter.section_ids.map((sectionId) => plan.sections.find((section) => section.id === sectionId));
    const currentContinuity = JSON.stringify([chapter.viewport, chapter.account_ref, chapter.project_state_ref, chapter.style_version, chapter.cursor_continuity_key]);
    if (chapter.project_id !== plan.project_id || chapter.order !== index || chapter.start_ms !== chapterSections[0]?.start_ms || chapter.end_ms !== chapterSections.at(-1)?.end_ms || chapter.end_ms <= chapter.start_ms || currentContinuity !== continuity) context.addIssue({ code: "custom", message: "Capture chapter lineage, timing, or continuity is invalid" });
  });
  let intervalEnd = 0;
  plan.condensed_intervals.forEach((interval) => {
    if (interval.start_ms < intervalEnd || interval.end_ms > plan.duration_ms) context.addIssue({ code: "custom", message: "Condensed intervals must be ordered within the video" });
    intervalEnd = interval.end_ms;
  });
});
export const longFormAdkStepSchema = z.object({
  id, run_id: id, project_id: id, kind: z.enum(["narrative", "template", "attention", "style"]),
  status: z.enum(["succeeded", "failed"]), tool_call_count: z.number().int().min(0).max(1),
  artifact_refs: z.array(id).max(160).default([]), output_sha256: z.string().regex(/^[a-f0-9]{64}$/).nullable().default(null),
  created_at: z.string().datetime({ offset: true }),
}).strict();
export const longFormAdkRunSchema = z.object({
  id, project_id: id, job_id: id, parent_motion_run_id: id, session_id: id,
  agent_name: z.literal("demodirector_longform_director"), model: id,
  status: z.enum(["running", "succeeded", "failed"]), runner_completed: z.boolean(),
  workflow_runs: z.number().int().min(0).max(1), step_ids: z.array(id).max(4),
  output_plan_id: id.nullable().default(null), started_at: z.string().datetime({ offset: true }),
  completed_at: z.string().datetime({ offset: true }).nullable().default(null),
  elapsed_ms: z.number().int().nonnegative().nullable().default(null), message: z.string().min(1).max(300),
}).strict();
export type LongFormVideoPlan = z.infer<typeof longFormVideoPlanSchema>;
export type ChapterCheckpoint = z.infer<typeof chapterCheckpointSchema>;
